Sage on nixos

timokau opened the pull request:
###### Motivation for this change

The current sage package rebuilds all its dependencies with its own build system. Thats a mess:

- It takes a long time
- Its basically 50% luck it works at all. Probably lots of failing tests too, I never tried.
- Its hard to patch
- It has to re-invent some nixpkgs features to patch the build system
- Its hard to integrate with other sofware (own python etc.)
- Its big

So this PR builds all the dependencies with nix instead. All the tests are passing.

It depends on (and includes) #39555, ~#39405~, ~#38794~ (without tests), ~#38787~ and the ~mpfr update in staging~. #39405 is responsible for most of the diff.

~It is not quite ready to merge yet, because it updates to sage 8.2.rc4 (last rc) which is not released yet and I still need to fix some FIXMEs, do some TODOs, document patches and remove redundant dependencies.~

I'm creating this WIP PR because 8.2 is about to be released and I don't want anybody else wasting time updating sage.

Testing this will take a lot of rebuilding because of the mpfr update (stdenv) and the python patch. ~**If you want to review, your time is probably best spend waiting until I'm finished**~. If you want to give feedback already, that is of course welcome.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

timokau committed:
openblas: expose version attribute

timokau committed:
libgap: switch to upstream url

timokau committed:
fetchurl: add sageupstream mirror

timokau committed:
combinatorial_designs: init at 20140630

timokau committed:
graphs: init at 20161026

timokau committed:
pari-galdata: init at 20080411

timokau committed:
pari-seadata-small: init at 20090618

timokau committed:
polytopes_db: init at 20170220

timokau committed:
conway_polynomials: init at 0.5

timokau committed:
elliptic_curves: init at 0.8

timokau committed:
gap-libgap-compatible: init at 4r8p6

timokau committed:
sage: 8.1 -> 8.2, use system packages

timokau committed:
tachyon: add armv7l support

timokau committed:
ecl_16_1_2: backport getcwd fix

timokau committed:
sympow: actually install datafiles

7c6f434c commented:
1. Impressive.

2. I think among the things to do before removing [WIP] is making sure you don't commit any clearly temporary files.

timokau commented:
Thanks :)

Oh right the `.orig` files. I put the PR up in a bit of a hurry and missed those. Removed.

timokau commented:
I'm not sure about the ofBorg error, it appears to have something to do with the node packages. I'm not sure about those anyways, I'm still hoping for some feedback from someone who knows more about node packaging.

timokau committed:
symmetrica: add patches

timokau committed:
sympy: backport upstream fixes

FRidh reviewed (changes requested):
Thanks @timokau , it is good to see this all together. 

I went through it, and then the changes look good to me, although I am still of the opinion we should not be adding old and unmaintained packages to Nixpkgs. Therefore, if these dependencies are really needed, I suggest you move them under the scope of `sage` and out of `python-packages.nix` and `python-modules/`.

7c6f434c commented:
@FRidh I am not sure if it is fully correct to call such packages unmaintained, as the Sage team is clearly capable and quite likely to perform maintenance on these packages if unavoidable (of course they can also drop the dependency, in which case we would find out the packages were indeed unmaintained along the way…)

On the other hand, maybe they are not worth putting in the global Python namespace, just as `liborcus` for LibreOffice lived together with it because nothing else was likely to ever use it.

timokau committed:
linbox: remove liblapack dependency

Not actually needed by any package and adds and indirect dependency to
atlas.

timokau commented:
Can somebody (@7c6f434c, @FRidh?) help me figure out how to avoid all the `inherit xyz` in `default.nix`? I think I need some of `newScope`, `makeScope`, `import pkgs.path { overlays = [ (self: super: {...})] }` and `python.pkgs.override`. But whatever I try, it either eats all my RAM in endless recursion, doesn't recognize the overrides or doesn't allow me to define new packages.

Could someone give me a minimal example how to create a new scope that overrides one python package and one regular pacakge? Something like:

```
pkgs = nixpkgs.giveMeNewScope (self: super: {
    glpk = super.glpk.overrideDerivation (...); # regular package
    numpy = super.python2.pkgs.numpy.override {}; # python package
    fantasy-py = self.python2.callPackage ./fantasy-py.nix {}; # new python package
});
```

7c6f434c commented:
I am not sure about proper structure of overrides for Nixpkgs nowadays.

Reimporting nixpkgs replacing user overlays with a sage-related one may
also be counterintuitive.

I am pretty sure you want something like:

self: super: rec {
  pkg1 = super.pkg1.override {…};

  callPackage = super.newScope self;

  pkg2 = callPackage … { };
}

timokau commented:
That defines a function right? But where do I call that function / give it as an argument?

7c6f434c commented:
I am not sure — reimporting with overlays sound dirty, packageOverrides are deprecated…

timokau commented:
Yeah reimporting with overlays is what the manual suggested, but I think thats probably a different use-case. Do you know who to ask?

7c6f434c commented:
Oh? Maybe look up in git who added that advice to ask if it applies in the case you want…

timokau commented:
That would be @grahamc in `doc/functions.xml` and @nbp in `doc/language-frameworks/python.section.md`. Any help would be much appreciated.

timokau commented:
Alright I think this is ready for review now. The only remaining blocker for merge is #39555. Once thats merged and through staging I will rebase this PR.

If I can clarify anything, let me know.

teto commented:
thanks that's an awesome contribution. I wanted to use sage at some point, had some problems, wanted to fix it but gave up when facing the huge task (plus sage developers don't seem too receptive about packaging concerns). I am no maintainer but I believe merging the new packages in one or several PRs would make things easier. I've seen that a maintainer didn't want merging one (old) python package but I would advocate that sage is worth it ?

timokau commented:
> thanks that's an awesome contribution.

You're welcome :)

> I wanted to use sage at some point, had some problems, wanted to fix it but gave up when facing the huge task (plus sage developers don't seem too receptive about packaging concerns).

There is *some* work ongoing upstream, but its not as much of an priority as I would like, true. I have also suggested all my relevant changes upstream[1] and at least none of them have been flat out rejected yet. So hopefully it will get better in the future.

> I am no maintainer but I believe merging the new packages in one or several PRs would make things easier. I've seen that a maintainer didn't want merging one (old) python package but I would advocate that sage is worth it ?

I have already split most new packages, package fixes and package updates in seperate PRs (lots of them). They've all been merged by now. The only things remaining are the pythong packages @FRidh doesn't want in the main tree, packages that don't make any sense without sage (like pybrial) and sage itself.

I think more or less the only thing blocking this PR is #39555, which is itself only blocked on the staging branch being a mess right now (not sure if thats still the case).

So once staging is cleaned up, #39555 can be merged into staging. Then we have to wait for staging to be merged into master and then this can be merged.


[1] https://trac.sagemath.org/query?status=closed&status=needs_info&status=needs_review&status=needs_work&status=new&status=positive_review&reporter=gh-timokau&order=priority

timokau committed:
pythonPackages.flask-autoindex: 0.6 -> 2018-06-28

Fixes the build with current flask, which deprecated the old import
syntax.

timokau committed:
pythonPackages.flask-silk: 0.2 -> 2018-06-28

Fixes the build with current flask, which deprecated the old import
syntax.

timokau committed:
cysignals: add pari support

timokau commented:
Finally all the dependencies are in master. I've cleaned up the history and rebased on nixpkgs-unstable for easier testing.

I think this is ready for test/merge.

@GrahamcOfBorg build sage.lib

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=367d8e1f-8dad-4b11-b102-a10a6568e3f3)

The following builds were skipped because they don't evaluate on x86_64-darwin: sage.lib

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=4744a8fe-f8ab-4cc2-83b6-670b95f3cfe9)

Attempted: sage.lib

<details><summary>Partial log (click to expand)</summary><p>

```
make[1]: Leaving directory '/build/source'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/v92icw9c0xqhml86n5xl1yj7rsdd4q9q-pynac-0.7.16
shrinking /nix/store/v92icw9c0xqhml86n5xl1yj7rsdd4q9q-pynac-0.7.16/lib/libpynac.so.17.1.0
strip is /nix/store/0pjsgkxz0rp5baycq5sp2s72lrr5q9sg-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/v92icw9c0xqhml86n5xl1yj7rsdd4q9q-pynac-0.7.16/lib
patching script interpreter paths in /nix/store/v92icw9c0xqhml86n5xl1yj7rsdd4q9q-pynac-0.7.16
checking for references to /build in /nix/store/v92icw9c0xqhml86n5xl1yj7rsdd4q9q-pynac-0.7.16...
cannot build derivation '/nix/store/82rrk0bx5hjv91hbjsj9qfbyyrvjnv82-python2.7-sagelib-8.2.drv': 6 dependencies couldn't be built
error: build of '/nix/store/82rrk0bx5hjv91hbjsj9qfbyyrvjnv82-python2.7-sagelib-8.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=40189b77-0d0b-4833-b246-27e43beb8adf)

Attempted: sage.lib

<details><summary>Partial log (click to expand)</summary><p>

```
100  1340  100  1340    0     0   1413      0 --:--:-- --:--:-- --:--:--  1413
building '/nix/store/c1614jmjiipk1aaswckmjwxhlsrchg3c-zero_division_error_formatting.patch.drv'...

trying https://git.sagemath.org/sage.git/patch/?h=f79070ddd09fa0ad6b340b097bd8d690a7aa35f0
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1631    0  1631    0     0   3294      0 --:--:-- --:--:-- --:--:--  3294
cannot build derivation '/nix/store/vsvv2dadn6mbdf4vzy9h4p09f59zksma-sage-src-8.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ivpj81ax61c5gvp31xj3icidsrg9bzil-python2.7-sagelib-8.2.drv': 4 dependencies couldn't be built
error: build of '/nix/store/ivpj81ax61c5gvp31xj3icidsrg9bzil-python2.7-sagelib-8.2.drv' failed
```
</p></details>

7c6f434c commented:
https://git.sagemath.org/sage.git/patch?id=d6ad2e2ccd7bce3c46cd2ae2f349941808d4d825 seems to be 404

FRidh commented on pkgs/applications/science/math/sage/default.nix:
```diff
@@ -1,230 +1,231 @@
+        });
+      };
+
+      sagenb = python.pkgs.callPackage ./sagenb.nix {
```
no `python.pkgs` but `self`.

FRidh commented on pkgs/applications/science/math/sage/default.nix:
```diff
@@ -1,230 +1,231 @@
+  });
+
+  # https://trac.sagemath.org/ticket/24824
+  glpk = nixpkgs.glpk.overrideDerivation (attrs: rec {
```
`overrideAttrs`

FRidh commented on pkgs/applications/science/math/sage/default.nix:
```diff
@@ -1,230 +1,231 @@
+  ecl = nixpkgs.ecl_16_1_2;
+
+  # sage currently uses an unreleased version of pari
+  pari = (nixpkgs.pari.override { withThread = false; }).overrideDerivation (attrs: rec {
```
`overrideAttrs`

FRidh commented on pkgs/applications/science/math/sage/pybrial.nix:
```diff
@@ -0,0 +1,29 @@
+# outside of sage anyways (as you could just directly depend on sage and use
+# it).
+buildPythonPackage rec {
+    pname = "pyBRiAl";
```
2 space indent

FRidh commented on pkgs/applications/science/math/sage/pybrial.nix:
```diff
@@ -0,0 +1,29 @@
+      sha256 = "0qy4cwy7qrk4zg151cmws5cglaa866z461cnj9wdnalabs7v7qbg";
+    };
+
+    preConfigure = "cd sage-brial";
```
is `sourceRoot` an option here?

timokau replied:
As a small rant: I don't like that you have to explicitly specify the name the "real" source root will have after unpacking when using `sourceRoot`. For example in this case I have to specify `sourceRoot = "source/sage-brial"`. I'd prefer if the logic that is used to determine the `sourceRoot` would always be executed and just rename the `sourceRoot` to something standard, like `src`. I think ArchLinux does it that way. Than it can just be optional weather or not to additionally `cd` into `src`.

But yes, `sourceRoot` is the better option here. I don't remember why I didn't use it, if I even had a conscious reason.

FRidh commented on pkgs/applications/science/math/sage/sage-env.nix:
```diff
@@ -0,0 +1,158 @@
+    "@sage-local@"
+    "@sage-local@/build"
+    pythonEnv
+    python3.buildEnv # empty env to clear PYTHONHOME
```
I don't follow this. Adding just `python3` is not sufficient?

timokau replied:
No, because then `wrapper.nix` won't be used and `PYTHONHOME` won't be re-set, see
https://github.com/NixOS/nixpkgs/blob/57bccb3cb8f6b5dea5c315fcaabdb992498b9314/pkgs/development/interpreters/python/wrapper.nix#L35

If I recall correctly that is important because sage calls the python3 binary from within python2. That means the python2 `PYTHONHOME` will be set and confuse python3.

The comment isn't very helpful though, I'll add more clarification.

FRidh commented on pkgs/applications/science/math/sage/sage-with-env.nix:
```diff
@@ -0,0 +1,129 @@
+
+  # remove python prefix, replace "-" in the name by "_", apply patch_names
+  # python2.7-some-pkg-1.0 -> some_pkg-1.0
+  pkg_to_spkg_name = pkg: patch_names: let
```
Most Python packages nowadays have a `pname` attribute. The plan is to make this mandatory.

FRidh commented on pkgs/applications/science/math/sage/sage-with-env.nix:
```diff
@@ -0,0 +1,129 @@
+
+
+  # return the names of all dependencies in the transitive closure
+  transitiveClosure = dep:
```
See also `python.pkgs.requiredPythonModules` although I don't think it is exactly what you are looking for.

timokau replied:
I don't 100% understand what that function does. But since I need it to work for *all* packages (non-python), unfortunately I don't think I can use that.

FRidh commented on pkgs/applications/science/math/sage/sagelib.nix:
```diff
@@ -0,0 +1,139 @@
+buildPythonPackage rec {
+  format = "other";
+  version = sage-src.version;
+  name = "sagelib-${version}";
```
`pname = "sagelib";` instead

FRidh reviewed (commented):
That's an impressive work! 

I've left some comments. As it doesn't touch any of the existing Python package sets I have no objections.

On a sidenote, considering you're using Python here a lot, and in various ways (packages, envs, ...) I am interested in knowing how you experienced the tooling. Any issues? Thinks you like to see different? Perhaps better to discuss is in another thread though.

(FYI, I'll be away on holidays and won't be able to give further feedback)

timokau commented:
> https://git.sagemath.org/sage.git/patch?id=d6ad2e2ccd7bce3c46cd2ae2f349941808d4d825 seems to be 404

I've fixed that link. I'm not entirely sure what caused that, but I've seen similar things happen before with the sage git server. I think it garbage-collects PR/ticket branches on force push or something.

Can we rely on hydras caching here or should I add all those patches to the `patches` folder? My goal is to ideally not have any patches in the `nixPatches` category for the next release (still in the process of upstreaming). But package-update backports will always be necessary.

@GrahamcOfBorg build sage.lib

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=02a2e9ab-ea56-4352-b1e2-cd975364cf39)

<details><summary>Partial log (click to expand)</summary><p>

```
Merge failed
```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=54fb10de-a191-4092-8837-ed9e5f0b98c9)

<details><summary>Partial log (click to expand)</summary><p>

```
Merge failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=9e5a8f40-3150-4316-bd64-4a54ddac4a40)

<details><summary>Partial log (click to expand)</summary><p>

```
Merge failed
```
</p></details>

timokau commented:
@FRidh thank you for your review, I know it is asking a lot to review such a big PR. I agree with all your comments and fixed the issues.

> On a sidenote, considering you're using Python here a lot, and in various ways (packages, envs, ...) I am interested in knowing how you experienced the tooling. Any issues? Thinks you like to see different? Perhaps better to discuss is in another thread though.

Some things took a while to figure out (especially the problem that makes the empty env necessary as described above). Overall I'm pretty happy though and much prefer the "hand-curated" approach to the completely auto-generated package sets.

In the beginning I used to just set `PYTHONPATH` manually, but the performance of that is atrocious (adding around a full second to startup time). So I'm glad that python envs exist.

Minor nitpicks/questsions:

- Its a bit irritating that some conventions are a tiny bit different than the rest of nixpkgs. Close enough to confuse, but not the same. `pname`, just using `python.pkgs` instead of passing everything explicitly as arguments and `overridePythonAttrs` come to mind. I kind of prefer the way things are done in python, but it does hurt consistency a bit.

- I don't quite understand why you do batch updates. Is there any practical reason for it? Why not handle python package updates like updates for any other package? For example the recent batch broke a few things for me and it is a bit hard to find the cause when everything is batched. Of course nobody can reasonably expect you to test every single update. So why not let the regular update bot create the updates and rely on the community to test and merge them?

- I don't understand why pypi (with `fetchFromPypi`) seems to be the preferred source for packages. Wouldn't it be preferrable to pull things directly from source control if available? That cuts out the middle-man and makes it easier to quickly test with an unreleased version.

All in all I'm very happy with it, thank you for maintaining :)

timokau commented:
@GrahamcOfBorg build sage.lib

Note that I'm still running the tests for the latest commit locally. Since the changes were pretty minor, I'm confident enough to push it already.

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=b092f393-8671-49be-b0eb-cfb2ef1c3a51)

The following builds were skipped because they don't evaluate on x86_64-darwin: sage.lib

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=6093dda6-32ad-4b54-bdae-fa05907873fb)

Attempted: sage.lib

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/src/bin/sage-upgrade: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/src/bin/sage-list-experimental: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/src/bin/sage-fix-pkg-checksums: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/src/bin/sage-list-optional: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/src/doc/common/update-python-inv.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/bootstrap: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
/nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2/sage: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash"
checking for references to /build in /nix/store/0qrzsvyxyfif38jsjfcsrnyh31ajf63h-sage-src-8.2...
cannot build derivation '/nix/store/h2n1zi1lqy494c82ih4smyfa8qi4b2db-python2.7-sagelib-8.2.drv': 3 dependencies couldn't be built
error: build of '/nix/store/h2n1zi1lqy494c82ih4smyfa8qi4b2db-python2.7-sagelib-8.2.drv' failed
```
</p></details>

timokau commented:
Seems like there is a timeout for the cython build. That's a shame, I'd love to see if ofBorg can build `sage.lib` (basically sage but no tests or wrappers).

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39981&attempt_id=d8bdd970-9467-4610-ab29-6959e15c9287)

Attempted: sage.lib

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2/lib/liblinboxsage.so.0.0.0
shrinking /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2/lib/liblinbox.so.0.0.0
gzipping man pages under /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2/share/man/
strip is /nix/store/0pjsgkxz0rp5baycq5sp2s72lrr5q9sg-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2/lib  /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2/bin
patching script interpreter paths in /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2
/nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2/bin/linbox-config: interpreter directive changed from " /bin/bash" to "/nix/store/p0vy17dp9jk2mvqsxsqnb14s3797lay7-bash-4.4-p23/bin/bash"
checking for references to /build in /nix/store/lw4qb9dam1gw6vhls6c2nbl9yfch0vmg-linbox-1.5.2...
cannot build derivation '/nix/store/3x6255f7545yx5djh0bzq7x5cxld8b1n-python2.7-sagelib-8.2.drv': 3 dependencies couldn't be built
error: build of '/nix/store/3x6255f7545yx5djh0bzq7x5cxld8b1n-python2.7-sagelib-8.2.drv' failed
```
</p></details>

timokau commented:
Does anybody know if the ofBorg evla failure is meaningful?

dotlambda commented:
@timokau Yes, the eval failure is because you removed `pkgs/development/python-modules/python-openid/default.nix` in ddd67df71d4e3304b094397cc8c1afbd13202d72. See https://github.com/NixOS/nixpkgs/commit/2e41891e6d595b9acf484d7d349b17a7b35b6923.

timokau commented:
Oh that must've been a mistake while fixing some merge conflict because I assumed that was my version of python-openid. I've rebased on latest master. Since python-openid for python2 is removed there, the eval should succeed now.

7c6f434c commented:
@timokau re: batch updates: when upstream of the language ecosystem has a package manager, it makes sense to track it via batch updates because the outcome is easier to reason about when comparing to the upstream situation.

timokau commented:
Why is that? Python packages can depend on specific version (ranges) and if a update builds in nixpkgs and doesn't break any reverse-dependencies, I see no reason not to include it right then.

dotlambda commented:
It's simply that just a small percentage of the python packages would be updated if we didn't have batch upgrades. And since the batch upgrades only happen on the unstable channel, people encountering broken builds should expect that this happens from time to time. Before a new release, we try to stabilize the python package set.
Btw, @FRidh proposed https://groups.google.com/forum/#!topic/nix-devel/koLvsCVjQcA, which would allow maintainers to opt out of batch upgrades.

timokau committed:
sage, arb: update sage dependencies arb and pynac

Removes the version pinning for arb and pynac by backporting the
upstream (sage) package upgrades. This necessitates a new patch for arb,
which was however already proposed and accepted upstream.

timokau committed:
sage: replace JUPYTER_PATH patch by upstream patch

The patch was replaced by a better upstream patch that doesn't mvoe the
jupyter dir to the user's home.

7c6f434c commented:
And a semi-related note: I think you should ask to get commit rights; while it might make sense to keep a Sage update in a PR while it is a WIP, and to discuss with @FRidh if you want a non-default version of a python package from a maintained branch, it is quite likely that reviewing your sage changes is useless because you have learnt about Sage much more than everyone else counted together. Also, I do think you have enough experience that your opinion about a PR «ah, it's just an uncontroversial update and it is ready to merge» is as good as anyone's.

teto commented:
Congrats ! thanks for the awesome work to all involved.

7c6f434c commented:
@rbvermaa @domenkozar I think it is a good idea to give @timokau write access.

7c6f434c commented:
@timokau in the meantime, do you have 2FA enabled on GitHub?

timokau commented:
Yes I have.

timokau commented:
@rbvermaa @domenkozar was this ever discussed?
